microsoft/terminal
# #459 IME UI does not follow the cursor in Windows Terminal

## Summary of the Pull Request
TerminalControl doesn't use any of the built in text input and edit controls provided by XAML
for text input, which means TermianlControl needs to communicate with the Text Services Framework (TSF)
in order to provide Input Method Editor (IME) support.  Just like the rest of Terminal we get
to take advantage of newer APIs ([Windows.UI.Text.Core namespace](https://docs.microsoft.com/en-us/uwp/api/windows.ui.text.core)) to provide support vs. the old TSF 1.0.

Windows.UI.Text.Core handles communication between a text edit control and the text services primarily
through a [CoretextEditContext](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Text.Core.CoreTextEditContext) object.

This change introduces a new UserControl TSFInputControl which is a custom EditControl similar to the [CustomEditControl sample](https://github.com/Microsoft/Windows-universal-samples/tree/master/Samples/CustomEditControl)

TSFInputControl is similar (overlay with IME text) to how old console (conimeinfo).  
 
## References


## PR Checklist
* [X] Closes #459, Closes #3641.


## Detailed Description of the Pull Request / Additional comments
TSFInputControl is a [Windows.UI.Xaml.Controls.UserControl](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.UserControl)

TSFInputControl contains a [Canvas](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.Canvas) control for absolution positioning a [TextBlock](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textblock) control within it's containing control (TerminalControl).

The TextBlock control is used for displaying candidate text from the IME.  When the user makes a choice in the IME the TextBlock is cleared and the text is written to the Terminal buffer like normal text.

TSFInputControl creates an instance of the CoreTextEditContext and attaches appropriate event handlers to CoreTextEditContext in order to interact with the IME.

A good write-up on how to interact with CoreTextEditContext can be found here: https://docs.microsoft.com/en-us/windows/uwp/design/input/custom-text-input

Text Updates:
Text updates from the IME come in on the TextUpdating event handler, text updates are stored in an internal buffer (_inputBuffer).

Completed Text:
Once a user selects a text in the IME, the CompositionCompleted handler is invoked.  The input buffer (_inputBuffer) is written to the Terminal buffer, _inputBuffer is cleared and Canvas and TextBlock controls are hidden until the user starts a composition session again.

Positioning:
Telling the IME where to properly position itself was the hardest part of this change.  The IME expects to know it's location in screen coordinates as supposed to client coordinates.  This is pretty easy if you are a pure UWP, but since we are hosted inside a XAMLIsland the client to screen coordinate translation is a little harder.  

How screen coordinate is calculated is by:
1. Obtaining the Window position in Screen coordinates.
2. Determining the Client coordinate of the cursor.
3. Converting the Client coordinate of the cursor to Screen coordinates.
4. Offsetting the X and Y coordinate of the cursor by the position of the TerminalControl within the window (tabs if present, margins, etc..).
5. Applying any scale factor of the display.

Once we have the right position in screen coordinates, this is supplied in the [LayoutBounds](https://docs.microsoft.com/en-us/uwp/api/windows.ui.text.core.coretextlayoutbounds) of the [CoreTextLayoutRequestedEventArgs](https://docs.microsoft.com/en-us/uwp/api/windows.ui.text.core.coretextlayoutrequestedeventargs) which let's the IME know where to position itself on the Screen.

Font Information/Cursor/Writing to Terminal:
3 events were added to the TSFInputControl to create a loosely coupled implementation between the TerminalControl and the TSFInputControl.  These events are used for obtaining Font information from the TerminalControl, getting the Cursor position and writing to the terminal buffer.

Known issues with this change:
- Width of TextBlock is hardcoded to 200 pixels and most likely should adjust to the available width of the current input line on the console.
- Entering text in the middle of an existing set of text has TextBlock render under existing text.  Current Console behavior here isn't good experience either (writes over text)
- Text input at edges of window is clipped versus wrapping around to next line.  This isn't any worse than the original command line, but Terminal should be better.

Future considerations:
Ideally, we'd be able to interact with the console buffer directly and replace characters as the user types. 


## Validation Steps Performed
General steps to try functionality
- Open Console
- Switch to Simplified Chinese (Shortcut: Windows+Spacebar)
- Switch to Chinese mode on language bar

Scenarios validated:
- As user types unformatted candidates appear on command line and IME renders in correct position under unformatted characters.
- User can dismiss IME and text doesn't appear on command line 
- Switch back to English mode, functions like normal
- New tab has proper behavior
- Switching between tabs has proper behavior
- Switching away from Terminal Window with IME present causes IME to disappear
